### ISSUE TITLE
Init fails on RaspberryPi 2

### ISSUE THREAD

**Issue by PTS93** [labatrockwell/ofxLibwebsockets#52]
The initialization just fails without any specific reason.
I tried it with 0.8.4 + master branch, 0.9.0 + master branch and 0.9.0 with 0.9.0 branch.
Also tried it with different server examples, all have the same issue.
The client examples gives a success for the init but are unable to connect.
Any ideas?

    [notice ] ofAppEGLWindow: setupPeripherals(): native event setup complete
    [ error ] [ofxLibwebsockets] libwebsockets init failed

**Comment by robotconscience** [comment]
Ah! Are you working with the master branch? Just realized I updated to the newest libwebsockets but only compiled for mac and win. Compiling libwebsockets is pretty simple; if you're able to do that and submit a PR I'd be really grateful!

1. Clone [libwebsockets](https://github.com/warmcat/libwebsockets)
2. Build LWS:
```
cd /path/to/libwebsockets
mkdir build
cd build
cmake .. -DLWS_IPV6=OFF
make
```
3. Move compiled lib into ofxLibwebsockets/libs/libwebsockets/lib/linuxarmv6l

**Comment by PTS93** [comment]
That seems to work, I made a PR. Though the RPI2 is ARMv7 not 6 so I added it to linuxarmv7l.

**Comment by PTS93** [comment]
While it now initializes it still seems to have some bugs.
When I use any of the server example I get:
`[warning] [ofxLibwebsockets] protocol is null`
The browser side (chrome) says:
`websocket connection CLOSED`
If I then reload the page the server is not reachable anymore from the browser and I get:

    [warning] [ofxLibwebsockets] Failed to send HTTP file /home/pi/of_v20150504_linuxarmv7l_release/addons/ofxLibwebsockets/example_server_sharedcanvas/bin/data/web/index.html for /index.html

**Comment by robotconscience** [comment]
Can you set ofLogLevel to OF_LOG_VERBOSE and post your trace here?

**Comment by PTS93** [comment]
When I load the page this gets printed. Though weirdly now it doesn't show that the connection closed, the web page just does nothing, no matter how often I reload. Without setting the log level it goes back to the previous behavior.
```
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_FILTER_NETWORK_CONNECTION
[verbose] [ofxLibwebsockets] LWS_CALLBACK_WSI_CREATE
[verbose] [ofxLibwebsockets] LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED
[verbose] [ofxLibwebsockets] LWS_CALLBACK_LOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_ADD_POLL_FD
[verbose] [ofxLibwebsockets] LWS_CALLBACK_UNLOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_FILTER_HTTP_CONNECTION
[verbose] [ofxLibwebsockets] LWS_CALLBACK_HTTP
[verbose] [ofxLibwebsockets] LWS_CALLBACK_HTTP_FILE_COMPLETION
[verbose] [ofxLibwebsockets] LWS_CALLBACK_CLOSED_HTTP
[verbose] [ofxLibwebsockets] LWS_CALLBACK_LOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_DEL_POLL_FD
[verbose] [ofxLibwebsockets] LWS_CALLBACK_UNLOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_CLOSED_HTTP
[verbose] [ofxLibwebsockets] LWS_CALLBACK_WSI_DESTROY
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_FILTER_NETWORK_CONNECTION
[verbose] [ofxLibwebsockets] LWS_CALLBACK_WSI_CREATE
[verbose] [ofxLibwebsockets] LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED
[verbose] [ofxLibwebsockets] LWS_CALLBACK_LOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_ADD_POLL_FD
[verbose] [ofxLibwebsockets] LWS_CALLBACK_UNLOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_FILTER_HTTP_CONNECTION
[verbose] [ofxLibwebsockets] LWS_CALLBACK_HTTP
[warning] [ofxLibwebsockets] Failed to send HTTP file /home/pi/of_v20150504_linuxarmv7l_release/addons/ofxLibwebsockets/example_server_sharedcanvas/bin/data/web/favicon.ico for /favicon.ico
[verbose] [ofxLibwebsockets] LWS_CALLBACK_LOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_DEL_POLL_FD
[verbose] [ofxLibwebsockets] LWS_CALLBACK_UNLOCK_POLL
[verbose] [ofxLibwebsockets] LWS_CALLBACK_CLOSED_HTTP
[verbose] [ofxLibwebsockets] LWS_CALLBACK_WSI_DESTROY
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
[verbose] [ofxLibwebsockets] LWS_CALLBACK_GET_THREAD_ID
```

**Comment by robotconscience** [comment]
Hm! Ok, try this:
In Reactor.cpp (ofxLibwebsockets/libs/ofxLibwebsockets/src), add these lines:
```
// top of file, after #include Reactor.h
#include "ofxLibwebsockets/Util.h"

//line 114, AFTER
/*
            if (conn == NULL){
                ofLog(OF_LOG_WARNING, "[ofxLibwebsockets] connection is null ");
            } else {
                ofLog(OF_LOG_WARNING, "[ofxLibwebsockets] protocol is null");
            }
*/
// ADD:
ofLog( OF_LOG_WARNING, "[ofxLibwebsockets] "+ getCallbackReason(reason) );
```

Now see what LWS event is occurring at the null protocol.

**Comment by PTS93** [comment]
```
[warning] [ofxLibwebsockets] protocol is null
[warning] [ofxLibwebsockets] LWS_CALLBACK_ESTABLISHED
[warning] [ofxLibwebsockets] protocol is null
[warning] [ofxLibwebsockets] LWS_CALLBACK_SERVER_WRITEABLE
[warning] [ofxLibwebsockets] protocol is null
[warning] [ofxLibwebsockets] LWS_CALLBACK_CLOSED
[warning] [ofxLibwebsockets] Failed to send HTTP file /home/pi/of_v20150504_linuxarmv7l_release/addons/ofxLibwebsockets/example_server_sharedcanvas/bin/data/web/favicon.ico for /favicon.ico
[warning] [ofxLibwebsockets] Failed to send HTTP file /home/pi/of_v20150504_linuxarmv7l_release/addons/ofxLibwebsockets/example_server_sharedcanvas/bin/data/web/index.html for /index.html
```

**Comment by robotconscience** [comment]
Hm! Can you try two things for me?
1. Test the libwebsocket examples on your RPi
2. Change ~line 110 to this:
```
            if (conn == NULL){
                ofLog(OF_LOG_WARNING, "[ofxLibwebsockets] connection is null ");
            } else if (conn->ws == NULL) {
                ofLog(OF_LOG_WARNING, "[ofxLibwebsockets] websocket is null");
            } else {
                ofLog(OF_LOG_WARNING, "[ofxLibwebsockets] protocol is null");
            }
```
and print what's happening?

It's sometimes OK that the protocol is null, but something weird is happening with the websocket creation, I think.

**Comment by PTS93** [comment]
The libwebsocket examples are working fine.
Replacing that code results in a complete crash of the app ending with this:
```
[warning] [ofxLibwebsockets] protocol is null
[warning] [ofxLibwebsockets] LWS_CALLBACK_ESTABLISHED
Segmentation fault
```

**Comment by robotconscience** [comment]
Oof. That doesn't make a ton of sense, but you can try changing
```
} else if (conn->ws == NULL) {
```
to
```
} else if (conn != NULL && conn->ws == NULL) {
```

Last test: can you try creating a separate html app and connecting to the websocket server? You can even test in Chrome directly through the debugger console by pasting
```
var ws = new WebSocket("ws://locahost:9093"); 
ws.onmessage = function(m){ console.log(m); } 
ws.onclose = function(){ console.log("closed");}
```
when your app is running...

**Comment by PTS93** [comment]
Unfortunately the same behavior with the snippet, with the only exception that I of course don't get any errors about not being able to send the index.html.
Changing the if statement doesn't change anything either.

**Comment by robotconscience** [comment]
OK! The if statement won't fix anything, just trying to get to the bottom of your problem. I had suspected it was with libwebsocket's http server but now I don't think so. I unfortunately do not have an RPi 1 or 2 to test with and can't replicate the errors on OS X. 

Something's happening internally with libwebsockets that's different on RPi.

Last thing to try is building from the exact commit I used:
https://github.com/warmcat/libwebsockets/commit/c67dace4be3309acffbad121e290d685308f0b86

**Comment by PTS93** [comment]
Tried to do that aswell but same outcome, very weird.

**Comment by robotconscience** [comment]
Can you try using the latest build of the library AND swapping out 
ofxLibwebsockets/libs/libwebsockets/include/libwebsockets.h for the one that compiles with the lib?

I'm still pretty sure it's an internal libwebsockets error, but all I can do is suggest fixes! Apologies, all RPi stuff is contributed and not maintained by me.

**Comment by PTS93** [comment]
That works! I had to fix a function call in reactor.cpp but that was about it.
Though I also had to include the lws_config.h that got build. The only libwebsockets.h I could find was in /libwebsockets/lib/ not in the build directory.

I'll make a pull request. I also changed the sharedcanvas example though you might not want to merge that ofAppGlutWindow doesn't seem to be supported on the Pi. Wouldn't link properly if I left that in, maybe I'm just missing a library.
